include/flexible.hxx

```diff
@@ -0,0 +1,464 @@
+      if (fields[i]){
+	((F*)fields[i])->applyBoundary(init);
+      }
+    }
+  }
+  virtual void applyTDerivBoundary() {
+    throw BoutException("Not implemented");
+  };
+  void allocate(){
+    ((F*)fields[mainid])->allocate();
+  }
+  Flexible<F>& operator*=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator*=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator*=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator*=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator*=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator*=(rhs);
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator/=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator/=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator/=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator/=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator/=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator/=(rhs);
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator+=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator+=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator+=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator+=(rhs);
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(const Field3D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator-=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(const Field2D & rhs) {
+    if (mainid == getId(rhs.getLocation())){
+      ((F*)fields[mainid])->operator-=(rhs);
+    } else {
+      throw BoutException("Not yet implemtented!");
+    }
+    clean(false);
+    return *this;
+  };
+  Flexible<F>& operator-=(BoutReal rhs) {
+    ((F*)fields[mainid])->operator-=(rhs);
+    clean(false);
+    return *this;
+  };
+
+private:
+  // Helper function to get index of location.
+  uint getId(CELL_LOC loc_){
+    uint loc = static_cast<uint>(loc_)-1;
```

Comment:
When `loc_` is 0, this will wrap-around

Reply:
Should I add some extra checks to prevent this underflow - or are we happy if I test in the line below for underflow?